[thennant] iP

thennant opened the pull request:
# C.R.Y.S.T.A.L
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))
### Creative. Reader for. Setting. Tasks. And. List
C.R.Y.S.T.A.L is your personal assistant for creating your personal task list. It's, 
- Easy to use
- ~~fast~~ **Super Fast**
- Text-based

All you need to do is,
1. Download it from [here](https://github.com/thennant/ip/releases/tag/A-Jar)
2. run java -jar crystal.jar
3. add your task
4. Sit back and watch your personal list generate :smiley:

and its **FREE!**

Current features include:
- [x] Marking tasks
- [x] Unmarking tasks
- [x] Handling various tasks such as Todos, Deadline and Events
- [ ] Reminders (coming soon)
##
#### If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

thennant committed:
created skeletal version of crystal

thennant committed:
added new features "add and list" to Crystal

thennant committed:
added mark and unmark features to Crystal

thennant committed:
added todo, event, deadline features to Crystal

thennant committed:
added Crystal Exception class and handle empty todo description and unrecognised term exceptions

thennant committed:
added delete feature to Crystal

thennant committed:
Change the code for string formatting of event and deadline

thennant committed:
done automated Text-UI testing

thennant committed:
Created save and load data
Made task class abstract
Remove uncessary methods from Todo, Event, Deadline class

thennant committed:
taught Crystal how to recognise dates and time

thennant committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Crystal.java
Settled merged conflicts

thennant committed:
Added Ui, Storage, Parser, Tasklist and Command classes

thennant committed:
Divided classes into packages

thennant committed:
Merge branch 'add-gradle-support'

thennant committed:
added JUnit test, change the deadline and event classes date time inputs and outputs

thennant committed:
Added java docs to all classes

thennant committed:
Tweak the code to comply with the coding standard

thennant committed:
added find feature for users to find a task by searching for a keyword.

thennant committed:
Merge branch 'branch-A-JavaDoc'

thennant committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/crystal/CrystalException.java
#	src/main/java/crystal/Storage.java
#	src/main/java/crystal/Ui.java
#	src/main/java/crystal/command/DeadlineCommand.java
#	src/main/java/crystal/command/TodoCommand.java
#	src/main/java/crystal/task/Deadline.java
#	src/test/java/crystal/task/EventTest.java

thennant committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/crystal/Parser.java
#	src/main/java/crystal/task/Task.java

laihuiqi commented on src/main/java/crystal/Parser.java:
```diff
@@ -0,0 +1,74 @@
+     * @param command User command
+     * @throws CrystalException When user command is not recognised
+     */
+    public static Command parse(String command) throws CrystalException {
```
Your `parse` method seems too long to me... Maybe you would like to try splitting them out w.r.t. the `command`? As suggestion, maybe a `handleCommand` or `parseCommand` private method can be introduced to increase the readability of this method?

thennant replied:
Ok, i will try it out!

laihuiqi commented on src/main/java/crystal/Ui.java:
```diff
@@ -0,0 +1,179 @@
+     *
+     */
+    public void printUnmark(TaskList task, int num) {
+        Task unmarkTask = task.get(num-1);
+        System.out.println(" ____________________________________________________________");
+        System.out.println("Alright, I've marked this task as not done: ");
+        unmarkTask.isDone = false;
+        System.out.println(unmarkTask.toString());
+        System.out.println(" ____________________________________________________________");
```
I like your output ui style !!!  It looks very neat and crystal clear ! :star_struck::star_struck::star_struck:

thennant replied:
Thank you!

laihuiqi commented on src/main/java/crystal/Storage.java:
```diff
@@ -0,0 +1,200 @@
+                    str.append(stringReplaceT + "\n");
+                }
+
+            } else if (tasks.get(i).toPrint().startsWith("[D]")) {
+                String stringReplaceD = tasks.get(i).toPrint().replace("[D]", "D ");
+                if (tasks.get(i).toPrint().contains("[X]")) {
+                    stringReplaceD = stringReplaceD.replace("[X]", "| 0 |");
+                    stringReplaceD = stringReplaceD.replace("(by:", "|");
+                    stringReplaceD = stringReplaceD.replace(")", "");
+                    str.append(stringReplaceD + "\n");
+                } else {
+                    stringReplaceD = stringReplaceD.replace("[ ]", "| 1 |");
+                    stringReplaceD = stringReplaceD.replace("(by:", "|");
+                    stringReplaceD = stringReplaceD.replace(")", "");
+                    str.append(stringReplaceD + "\n");
+                }
```
Seems that this actual string to stored data transformation process is long and repeating... I have noticed that you have created classes for task objects, so, maybe you would like to try including them as class methods, in the aspect of better readability and ease of reuse? :thinking:

thennant replied:
Ok, i will try that

laihuiqi commented on src/test/java/crystal/command/DeadlineCommandTest.java:
```diff
@@ -0,0 +1,21 @@
+ * Represents the DeadlineCommandTest class.
+ *
+ */
+public class DeadlineCommandTest {
```
Sufficient test cases are provided for every task objects classes !! :+1::thumbsup:

laihuiqi commented on src/main/java/crystal/Storage.java:
```diff
@@ -0,0 +1,200 @@
+
+                } else if (nextInput.startsWith("E")) {
+                    if (nextInput.contains("| 0 |")) {
+                        String description = nextInput.replace("E | 0 |", "");
+                        int index = description.lastIndexOf("|");
+                        int index2 = description.lastIndexOf("-");
+                        String time = description.substring(description.lastIndexOf("|") + 1);
+                        int index3 = time.lastIndexOf(" - ");
+                        time = time.replace(time.substring(index3), "");
+                        String endTime = description.substring(description.lastIndexOf(" - ") + 3);
+                        description = description.replace(description.substring(index), "");
+                        Task newTask = new Event(description.trim(), time.trim(), endTime.trim());
+                        newTask.isDone = true;
+                        temp.add(newTask);
```
Would you like to add some line spaces to categorize some of the variables?

laihuiqi reviewed (commented):
Nice to meet you !!! :wave:
Overall, your code is well organized w.r.t. the OOP-principles. Each relevant classes are also packed well in the same packaged. However, some of your code looks lengthy so it may affect the readability a little bit and I have given some tiny suggestions on it. 
Really like your CLI's output UI! :laughing:

thennant committed:
Added a GUI to Crystal.

anshumaantgi commented on src/main/java/crystal/CrystalException.java:
```diff
@@ -0,0 +1,19 @@
+ * Represents the CrystalException class.
+ *
+ */
+public class CrystalException extends Exception {
```
I like how you defined a single exception for your project. How abt putting the Excpetion in a folder, so in the future, if you make more exceptions it's easier to add and also help you're testing?

anshumaantgi commented on src/main/java/crystal/MainWindow.java:
```diff
@@ -0,0 +1,56 @@
+/**
+ * Controller for MainWindow. Provides the layout for the other controls.
+ */
+public class MainWindow extends AnchorPane {
```
I like how you have divided the UI into classes too , maybe putting them in folders as well would make you're directory cleaner .

anshumaantgi commented on src/main/java/crystal/Parser.java:
```diff
@@ -0,0 +1,75 @@
+     * @param command User command
+     * @throws CrystalException When user command is not recognised
+     */
+    public static Command parse(String command) throws CrystalException {
```
Also , Have you thought of using switch-case for you're parser? It makes the code so much more neater

anshumaantgi commented on src/main/java/crystal/Parser.java:
```diff
@@ -0,0 +1,75 @@
+            stringNoDeadline = stringNoDeadline.substring(index + 3); //date/time for by portion
+            return new DeadlineCommand(description, stringNoDeadline);
+        } else if (command.contains("event")) {
+            String stringNoEvent = command.replace("event", "");
```
All these functionality in the commands block can maybe be abstracted into functions. Good job on cleaning your input string , looks like a good defensive programming practice

anshumaantgi commented on src/main/java/crystal/Storage.java:
```diff
@@ -0,0 +1,200 @@
+        StringBuilder str = new StringBuilder();
+        for (int i = 0; i < tasks.size(); i++) {
+
+            if (tasks.get(i).toString().startsWith("[T]")) {
```
Think abt abstracting the task representation in the storage file, inside the Task object, this way, if you have more tasks in the future, you do not need to touch the storage.java file.

anshumaantgi commented on src/test/java/crystal/task/DeadlineTest.java:
```diff
@@ -0,0 +1,23 @@
+ * Represents the DeadlineTest class.
+ *
+ */
+public class DeadlineTest {
```
Maybe try using Test names  according to convention `featureUnderTest_testScenario_expectedBehavior()`

anshumaantgi reviewed (commented):
Overall a very creative and solid IP. I can clearly see the amount of work you have put into it and Kudos to you for keeping up with the deadlines. Just need to maybe try abstracting some code and files , to make it look neater , otherwise looks good to me.

thennant committed:
TaskList class: Added assertion to get() and remove() methods
Assert is added to check for negative index

Event class: Added assert to Event constructor
Assert is added to check that the to date is not before the the from date

Task class: Added assert to getDescription() method
Assert is added to check that description of task is not empty

thennant committed:
Add Checkstyle

thennant committed:
Merge branch 'branch-A-CheckStyle'

thennant committed:
* Added checkstyle
* Refactored parser class code
* Improve code quality for all classes
* Removed unnecessary code

thennant committed:
Merge pull request #1 from thennant/branch-A-Assertions

TaskList class: Added assertion to get() and remove() methods

thennant committed:
Merge branch 'branch-A-Assertions'

thennant committed:
Merge pull request #2 from thennant/branch-A-CodeQuality

Branch a code quality

thennant committed:
Merge branch 'branch-A-CodeQuality'

thennant committed:
Merge branch 'master' of https://github.com/thennant/ip

* 'master' of https://github.com/thennant/ip:

thennant committed:
Add Priority Command

Priority Command was added as an extension feature to allow users to set priority tags and level to the task of their choice

Priority Command class was made to implement this feature

It is implemented as a class as the user is expected  to input the priority command to set the tag to the specific task

thennant committed:
Merge branch 'branch-BCD-Extension'

thennant committed:
Add user guide and representative image

Changed profile picture of crystal and user

thennant committed:
Add Ui image and edit section hyperlinks

thennant committed:
Update README.md

thennant committed:
Merge branch 'master' of https://github.com/thennant/ip

* 'master' of https://github.com/thennant/ip:
  Update README.md

thennant committed:
Add exit feature to Main Window

Exit feature was added so that when bye command is entered, the window will close after 1.5 seconds

Updated the user guide

thennant committed:
Storage class code is too long and reads the text file as a string.

The user cannot overwrrite the old data.

Storage class code is reformatted to be shorter and read the data dynamically.
This allows the user to overwrite the old data

ReadME md is also updated
